From 54bfe201cc5f7014c2227f69a3e5fa7a6aff72da Mon Sep 17 00:00:00 2001
From: Kaan Alsan <alsan@chromium.org>
Date: Thu, 7 May 2026 04:36:33 -0700
Subject: [PATCH] [Vertical Tabs] Don't end SystemDnD for "capture lost" events

Some window managers don't send "exit" events when the drag detaches
from a tabstrip, which causes a crash when the drag attaches to a new
tab strip, since that sends the "capture lost" event to the original one
(see bug for an example of this).

This CL updates the vertical tab strip to match the behavior of the
horizontal tab strip, which does not cancel SystemDnD when mouse capture
is lost.

Fixed: 505023370
Change-Id: Ica9a7332b8a05abfa1845bdc87c460cbf9f8206e
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7821001
Reviewed-by: Vince Lugli <lugli@google.com>
Commit-Queue: Kaan Alsan <alsan@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1626845}
---
 .../ui/views/tabs/vertical/vertical_tab_drag_handler.cc   | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc b/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
index 306778f49fc894..737072a5bd3380 100644
--- a/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_tab_drag_handler.cc
@@ -328,7 +328,13 @@ bool VerticalTabDragHandlerImpl::ContinueDrag(views::View& event_source_view,
 }
 
 void VerticalTabDragHandlerImpl::EndDrag(EndDragReason reason) {
-  if (TabDragController::IsSystemDnDSessionRunning()) {
+  // Note: we avoid ending the SystemDnD if the `EndDragReason` is "capture
+  // lost" because some window managers on Wayland don't send exit events to tab
+  // strips, which would cause this to incorrectly end the SystemDnD while the
+  // drag is attaching to a new tab strip. See http://crbug.com/505023370 for an
+  // example of this.
+  if (TabDragController::IsSystemDnDSessionRunning() &&
+      reason != EndDragReason::kCaptureLost) {
     TabDragController::OnSystemDnDEnded();
     return;
   }
